llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -22243,6 +22243,19 @@ SDValue DAGCombiner::visitEXTRACT_VECTOR_ELT(SDNode *N) {
   unsigned NumElts = VecVT.getVectorNumElements();
   unsigned VecEltBitWidth = VecVT.getScalarSizeInBits();
 
+  // See if the extracted element is constant, in which case fold it if its
+  // a legal fp immediate.
+  if (IndexC && ScalarVT.isFloatingPoint()) {
+    APInt EltMask = APInt::getOneBitSet(NumElts, IndexC->getZExtValue());
+    KnownBits KnownElt = DAG.computeKnownBits(VecOp, EltMask);
+    if (KnownElt.isConstant()) {
```

Comment:
Why is this fancy instead of just using dyn_cast<ConstantSDNode>?

Reply:
Its to handle bitcasts of vector data (e.g. <2 x i32>), which don't get folded away if we have multiple uses